### DISCUSSION TITLE
Possibility to set API_KEY in runtime

### DISCUSSION THREAD

**Discussion by dresl** [anymail/django-anymail#334] [Ideas]
Hi, it would be nice to set API key of email backend manually during runtime (from Admin), not just in settings. I have a project that needs to set multiple EmailBackends with multiple api keys. Would it be possible to add a `setter` like `set_api_key` to EmailBackend?

**Comment by dresl** [comment]
Resolved by adding `api_key` to kwargs as backend constructor.

**Comment by medmunds** [comment]
You can use [`django.core.mail.get_connection()`](https://docs.djangoproject.com/en/stable/topics/email/#django.core.mail.get_connection) to create an email backend instance with custom settings:

```python
from django.conf import settings
from django.core.mail import get_connection, EmailMessage

connection1 = get_connection(settings.EMAIL_BACKEND, api_key="api-key-1")
send_mail("Sent using api-key-1", ..., connection=connection1)

connection2 = get_connection(settings.EMAIL_BACKEND, api_key="api-key-2")
send_mail("Sent using api-key-2", ..., connection=connection2)

message = EmailMessage("Also sent using api-key-2", ..., connection=connection2)
message.send()
```

You can override almost all Anymail backend settings this way. So you could use `get_connection(..., api_url="https://europe-api-endpoint...")` to override the default API URL and send certain messages via your ESP's European servers.

In Anymail's docs, see [*Anymail settings reference*](https://anymail.dev/en/stable/installation/#anymail-settings-reference:~:text=override%20most%20settings%20on%20a%20per%2Dinstance%20basis) for more info on settings overrides, and [*Mixing email backends*](https://anymail.dev/en/stable/tips/multiple_backends/) for some specific `get_connection()` examples.
